### PR TITLE
Bring type safety to colour

### PR DESCRIPTION
**Pull request by duncanspumpkin**
I wanted to do something with imageId and that lead me on a winding path to wanting type safe colours. I'm not 100% on the names of everything so if anyone has better naming that would be great.

### CODE REVIEW

File: src/OpenLoco/Objects/WallObject.cpp [hunk 1]
```diff
@@ -26,10 +26,13 @@ namespace OpenLoco
     void WallObject::drawPreviewImage(Gfx::Context& context, const int16_t x, const int16_t y) const
     {
         auto image = sprite;
-        image = Gfx::recolour(image, Colour::mutedDarkRed);
         if (flags & (1 << 6))
         {
-            image |= (1 << 31) | (1 << 28);
+            image = Gfx::recolour2(sprite, Colour2::mutedDarkRed, Colour2::grey);
```
Comment (duncanspumpkin) [inline src/OpenLoco/Objects/WallObject.cpp:31]: incorrect second colour should be yellow

**Comment by duncanspumpkin** [comment]
For naming reference see https://github.com/OpenRCT2/OpenRCT2/blob/develop/src/openrct2/drawing/Drawing.h#L199 it follows the same general layout.

File: src/OpenLoco/GameCommands/GameCommands.h [hunk 2]
```diff
@@ -842,17 +842,17 @@ namespace OpenLoco::GameCommands
             , rotation(regs.dl)
             , type(regs.bh)
             , unk(regs.dh)
-            , primaryColour(regs.bp & 0xFF)
-            , secondaryColour((regs.bp >> 8) & 0xFF)
+            , primaryColour(static_cast<Colour>(regs.bp & 0x1F))
```
Comment (LeftofZen) [inline src/OpenLoco/GameCommands/GameCommands.h:845]: different values?

File: src/OpenLoco/Graphics/Colour.h [hunk 3]
```diff
@@ -131,9 +347,30 @@ namespace OpenLoco
...
+        constexpr PaletteIndex_t primaryRemap3 = 0xF6;
+        constexpr PaletteIndex_t primaryRemap4 = 0xF7;
+        constexpr PaletteIndex_t primaryRemap5 = 0xF8;
+        constexpr PaletteIndex_t primaryRemap6 = 0xF9;
+        constexpr PaletteIndex_t primaryRemap7 = 0xFA;
+        constexpr PaletteIndex_t primaryRemap8 = 0xFB;
+        constexpr PaletteIndex_t primaryRemap9 = 0xFC;
+        constexpr PaletteIndex_t primaryRemapA = 0xFD;
+        constexpr PaletteIndex_t primaryRemapB = 0xFE;
     }
```
Comment (LeftofZen) [inline src/OpenLoco/Graphics/Colour.h:375]: is there an 0xFF palette index?

File: src/OpenLoco/Objects/CompetitorObject.cpp [hunk 4]
```diff
@@ -13,9 +13,9 @@ namespace OpenLoco
     // 0x00434D5B
     void CompetitorObject::drawPreviewImage(Gfx::Context& context, const int16_t x, const int16_t y) const
     {
-        Gfx::drawRect(context, 0, 0, objectPreviewSize.width, objectPreviewSize.height, Colour::inset(Colour::mutedOrange));
+        Gfx::drawRect(context, 0, 0, objectPreviewSize.width, objectPreviewSize.height, AdvancedColour(Colour::mutedOrange).inset().u8());
 
-        auto image = Gfx::recolour(images[0], Colour::inset(Colour::mutedOrange));
+        auto image = Gfx::recolour(images[0], Colour::mutedSeaGreen);
```
Comment (LeftofZen) [inline src/OpenLoco/Objects/CompetitorObject.cpp:18]: `mutedOrange` became `mutedSeaGreen`, is that correct?

File: src/OpenLoco/Ui/Dropdown.cpp [hunk 5]
```diff
@@ -263,17 +263,17 @@ namespace OpenLoco::Ui::Dropdown
...
+                    if (!self->getColour(WindowColour::primary).isTranslucent())
                     {
-                        Gfx::drawRect(*context, x, y, _dropdownItemWidth - 1, 1, Colour::getShade(self->getColour(WindowColour::primary), 3));
-                        Gfx::drawRect(*context, x, y + 1, _dropdownItemWidth - 1, 1, Colour::getShade(self->getColour(WindowColour::primary), 7));
+                        Gfx::drawRect(*context, x, y, _dropdownItemWidth - 1, 1, Colours::getShade(self->getColour(WindowColour::primary).c(), 3));
+                        Gfx::drawRect(*context, x, y + 1, _dropdownItemWidth - 1, 1, Colours::getShade(self->getColour(WindowColour::primary).c(), 7));
                     }
                     else
                     {
-                        uint32_t colour = Colour::getTranslucent(Colour::opaque(self->getColour(WindowColour::primary))) | (1 << 25);
```
Comment (LeftofZen) [inline src/OpenLoco/Ui/Dropdown.cpp:273]: I like how the original code wanted the opaque colour, just to make it transparent immediately after

File: src/OpenLoco/Windows/CompanyList.cpp [hunk 6]
```diff
@@ -1026,7 +1026,7 @@ namespace OpenLoco::Ui::Windows::CompanyList
                     continue;
 
                 auto colour = _cargoLineColour[i];
-                colour = Colour::getShade(colour, 6);
+                auto pallete = Colours::getShade(colour, 6);
```
Comment (LeftofZen) [inline src/OpenLoco/Windows/CompanyList.cpp:1029]: wrong spelling for palette

**Comment by LeftofZen** [comment]
This is a big PR, I'll have to look at the rest tomorrow

File: src/OpenLoco/Graphics/Colour.h (hunk 3 above)
Comment (duncanspumpkin) [reply src/OpenLoco/Graphics/Colour.h:375]: Yup will add

**Review by LeftofZen — COMMENTED** [review]
few minor questions

File: src/OpenLoco/GameCommands/GameCommands.h (hunk 2 above)
Comment (duncanspumpkin) [reply src/OpenLoco/GameCommands/GameCommands.h:845]: Yeah i felt it was better to cap it to the correct range.

Comment (LeftofZen) [reply src/OpenLoco/GameCommands/GameCommands.h:845]: No worries, wasn't sure if it was a typo or not

File: src/OpenLoco/Ui/Dropdown.cpp (hunk 5 above)
Comment (duncanspumpkin) [reply src/OpenLoco/Ui/Dropdown.cpp:273]: Yeah technically `getTranslucent` is different to `Colour::translucent` as the first is used to get the palette that represents the translucent version of a colour and the other is just a flag to indicate that the colour is translucent.

File: src/OpenLoco/Objects/CompetitorObject.cpp (hunk 4 above)
Comment (duncanspumpkin) [reply src/OpenLoco/Objects/CompetitorObject.cpp:18]: Yes that is correct.

**Review by LeftofZen — APPROVED** [review]

**Pull request merged**